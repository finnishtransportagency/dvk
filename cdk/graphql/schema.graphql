schema {
  query: Query
  mutation: Mutation
}

type Query {
  fairwayCards(status: [Status!]): [FairwayCard!]!
  fairwayCard(id: String!, version: String!): FairwayCard
  fairwayCardVersion(id: String!, version: String!): FairwayCard
  fairwayCardLatest(id: String!): FairwayCard
  fairwayCardPublic(id: String!): FairwayCard
  fairwayCardPreview(id: String!, version: String!): FairwayCard
  safetyEquipmentFaults: [SafetyEquipmentFault!]!
  marineWarnings: [MarineWarning!]!
  currentUser: CurrentUser
  pilotPlaces: [PilotPlace!]!
  harbors: [Harbor!]!
  harbor(id: String!, version: String!): Harbor
  harborVersion(id: String!, version: String!): Harbor
  harborLatest(id: String!): Harbor
  harborPublic(id: String!): Harbor
  harborPreview(id: String!, version: String!): Harbor
  fairways: [Fairway!]!
  fairwayCardsAndHarbors: [FairwayCardOrHarbor!]!
  mareographs: [Mareograph!]!
}

type Mutation {
  saveFairwayCard(card: FairwayCardInput!): FairwayCard
  saveHarbor(harbor: HarborInput!): Harbor
  uploadPicture(picture: PictureUploadInput!): Boolean
}

input PictureUploadInput {
  cardId: String!
  id: String!
  base64Data: String!
  contentType: String!
}

enum Operation {
  CREATE
  UPDATE
}

enum OperationError {
  CARD_ID_MISSING
  CARD_ALREADY_EXIST
  CARD_NOT_EXIST
  HARBOR_ID_MISSING
  HARBOR_ALREADY_EXIST
  HARBOR_NOT_EXIST
  NO_PERMISSIONS
  INVALID_INPUT
}

input FairwayCardInput {
  id: String!
  version: String!
  operation: Operation!
  name: TextInput!
  n2000HeightSystem: Boolean!
  group: String!
  status: Status!
  fairwayIds: [Float!]!
  primaryFairwayId: [SelectedFairwayInput!]
  secondaryFairwayId: [SelectedFairwayInput!]
  additionalInfo: TextInput
  lineText: TextInput
  designSpeed: TextInput
  generalInfo: TextInput
  anchorage: TextInput
  navigationCondition: TextInput
  iceCondition: TextInput
  attention: TextInput
  speedLimit: TextInput
  visibility: TextInput
  vesselRecommendation: TextInput
  mareographs: [Int!]
  windRecommendation: TextInput
  trafficService: TrafficServiceInput
  harbors: [String!]
  pictures: [PictureInput!]
  pilotRoutes: [Int!]
  temporaryNotifications: [TemporaryNotificationInput!]
}

input PictureInput {
  id: String!
  sequenceNumber: Int
  orientation: Orientation!
  rotation: Float
  scaleWidth: String
  scaleLabel: String
  text: String
  lang: String
  groupId: Float
  modificationTimestamp: Float
  legendPosition: String
}

enum Orientation {
  PORTRAIT
  LANDSCAPE
}

input TrafficServiceInput {
  pilot: PilotInput
  vts: [VTSInput]
  tugs: [TugInput]
}

input TugInput {
  name: TextInput!
  phoneNumber: [String!]
  fax: String
  email: String
}

input VHFInput {
  name: TextInput
  channel: String!
}

input VTSInput {
  name: TextInput!
  phoneNumber: String
  email: [String]
  vhf: [VHFInput]
}

input PilotInput {
  email: String
  phoneNumber: String
  fax: String
  internet: String
  extraInfo: TextInput
  places: [PilotPlaceInput!]
}

input PilotPlaceInput {
  id: Float!
  pilotJourney: String
}

input HarborInput {
  id: String!
  version: String!
  n2000HeightSystem: Boolean!
  operation: Operation!
  status: Status!
  phoneNumber: [String]
  fax: String
  email: String
  internet: String
  name: TextInput!
  company: TextInput
  extraInfo: TextInput
  harborBasin: TextInput
  cargo: TextInput
  quays: [QuayInput]
  geometry: GeometryInput!
}

enum Status {
  DRAFT
  PUBLIC
  REMOVED
}

input GeometryInput {
  lat: String!
  lon: String!
}

input TextInput {
  fi: String!
  sv: String!
  en: String!
}

input QuayInput {
  name: TextInput
  length: String
  sections: [SectionInput]
  extraInfo: TextInput
  geometry: GeometryInput
}

input SectionInput {
  name: String
  depth: String
  geometry: GeometryInput
}

input TemporaryNotificationInput {
  content: TextInput!
  startDate: String!
  endDate: String
}

input SelectedFairwayInput {
  id: Int!
  sequenceNumber: Int!
}

type FairwayCardOrHarbor {
  id: String!
  version: String!
  type: ContentType!
  name: Text!
  group: String
  fairwayIds: [Int!]
  n2000HeightSystem: Boolean!
  status: Status!
  creator: String
  modifier: String
  modificationTimestamp: Float
  temporaryNotifications: [TemporaryNotification]
}

enum ContentType {
  CARD
  HARBOR
}

type CurrentUser {
  uid: String!
  name: String!
  roles: [String!]!
}

type FairwayCard {
  id: String!
  version: String!
  name: Text!
  n2000HeightSystem: Boolean!
  group: String
  status: Status
  creator: String
  creationTimestamp: Float
  modifier: String
  modificationTimestamp: Float
  fairwayIds: String
  fairways: [Fairway!]!
  additionalInfo: Text
  lineText: Text
  designSpeed: Text
  generalInfo: Text
  anchorage: Text
  navigationCondition: Text
  iceCondition: Text
  attention: Text
  speedLimit: Text
  visibility: Text
  vesselRecommendation: Text
  mareographs: [Mareograph!]
  windRecommendation: Text
  trafficService: TrafficService
  harbors: [Harbor!]
  pictures: [Picture!]
  pilotRoutes: [PilotRoute!]
  temporaryNotifications: [TemporaryNotification!]
}

type Picture {
  id: String!
  sequenceNumber: Int
  orientation: Orientation!
  rotation: Float
  text: String
  lang: String
  scaleLabel: String
  scaleWidth: String
  groupId: Float
  modificationTimestamp: Float
  legendPosition: String
}

type TrafficService {
  pilot: Pilot
  vts: [VTS]
  tugs: [Tug]
}

type Tug {
  name: Text
  phoneNumber: [String]
  fax: String
  email: String
}

type VHF {
  name: Text
  channel: Int
}

type VTS {
  name: Text
  phoneNumber: String
  email: [String]
  vhf: [VHF]
}

type Pilot {
  email: String
  phoneNumber: String
  fax: String
  internet: String
  extraInfo: Text
  places: [PilotPlace!]
}

type PilotPlace {
  id: Float!
  name: Text!
  pilotJourney: Float
  geometry: GeometryPoint!
}

type Harbor {
  id: String!
  version: String
  status: Status
  n2000HeightSystem: Boolean
  creator: String
  creationTimestamp: Float
  modifier: String
  modificationTimestamp: Float
  quays: [Quay]
  name: Text
  company: Text
  phoneNumber: [String]
  fax: String
  email: String
  internet: String
  extraInfo: Text
  harborBasin: Text
  cargo: Text
  geometry: GeometryPoint
}

type Quay {
  name: Text
  length: Float
  sections: [Section]
  extraInfo: Text
  geometry: GeometryPoint
}

type Section {
  name: String
  depth: Float
  geometry: GeometryPoint
}

type Fairway {
  id: Int!
  name: Text
  primary: Boolean
  primarySequenceNumber: Int
  secondary: Boolean
  secondarySequenceNumber: Int
  startText: String
  endText: String
  lightingCode: String
  lighting: Text
  typeCode: String
  type: Text
  owner: String
  area: Text
  sizing: Sizing
  classifications: [Classification!]
  sizingVessels: [SizingVessel!]
  navigationLines: [NavigationLine!]
  areas: [Area!]
  restrictionAreas: [RestrictionArea!]
  prohibitionAreas: [ProhibitionArea!]
  boardLines: [Boardline!]
  turningCircles: [TurningCircle!]
}

type Classification {
  type: String
  fairwayClassCode: String
  fairwayClass: Text
}

type AreaFairway {
  fairwayId: Int!
  status: Int
  line: Int
  sizingSpeed: Float
  sizingSpeed2: Float
  sequenceNumber: Int
}

type Area {
  id: Int!
  name: String
  depth: Float
  draft: Float
  referenceLevel: String
  n2000depth: Float
  n2000draft: Float
  n2000ReferenceLevel: String
  direction: Float
  journalNumber: String
  verificationDate: String
  owner: String
  additionalInformation: String
  typeCode: Int
  type: String
  notationCode: Float
  notation: String
  operationStatusCode: Float
  operationStatus: String
  operationTypeCode: Float
  operationType: String
  operationDirectionCode: Float
  operationDirection: String
  fairways: [AreaFairway!]
  geometry: GeometryPolygon
}

type RestrictionArea {
  id: Int!
  types: [RestrictionType!]
  value: Float
  presenter: String
  journalNumber: String
  verificationDate: String
  modificationDate: String
  startDate: String
  endDate: String
  status: String
  source: String
  location: String
  municipality: String
  exception: String
  geometry: GeometryPolygon
  fairways: [RestrictionFairway!]
}

type RestrictionType {
  code: String
  text: String
}

type RestrictionFairway {
  fairwayId: Int!
}

type ProhibitionArea {
  id: Int!
  extraInfo: Text
  fairway: ProhibitionFairway!
}

type ProhibitionFairway {
  fairwayId: Int!
  name: Text
}

type BoardlineFairway {
  fairwayId: Int!
}

type Boardline {
  id: Int!
  direction: Float
  geometry: GeometryLine
  fairways: [BoardlineFairway!]
}

type TurningCircleFairway {
  fairwayId: Int!
}

type TurningCircle {
  id: Int!
  geometry: GeometryLine
  diameter: Int
  fairways: [TurningCircleFairway!]
}

type SizingVessel {
  typeCode: String
  type: String
  length: Float
  width: Float
  draft: Float
  size: Float
  bodyFactor: Float
}

type Sizing {
  normalTurningCircle: Float
  minimumTurningCircle: Float
  normalWidth: Float
  minimumWidth: Float
  additionalInformation: String
  mareograph: String
  reserveWater: String
}

type NavigationLine {
  id: Int!
  depth: Float
  draft: Float
  referenceLevel: String
  n2000depth: Float
  n2000draft: Float
  n2000ReferenceLevel: String
  direction: Float
  length: Float
  owner: String
  verificationDate: String
  journalNumber: String
  additionalInformation: String
  type: String
  typeCode: String
  geometry: GeometryLine
  fairways: [NavigationLineFairway!]
}

type NavigationLineFairway {
  fairwayId: Int!
  status: Int
  line: Int
}

type SafetyEquipmentFault {
  id: Int!
  name: Text
  equipmentId: Int!
  typeCode: Int!
  type: Text
  recordTime: Float!
  geometry: GeometryPoint!
}

type GeometryPolygon {
  type: String
  coordinates: [[[Float]]]
}

type GeometryLine {
  type: String
  coordinates: [[Float]]
}

type GeometryPoint {
  type: String
  coordinates: [Float]
}

type Text {
  fi: String
  sv: String
  en: String
}

type MarineWarning {
  id: Int!
  number: Int!
  area: Text!
  type: Text!
  location: Text!
  description: Text!
  startDateTime: Float
  endDateTime: Float
  dateTime: Float!
  notifier: String!
  equipmentId: Float
  lineId: Float
  areaId: Float
}

type PilotRoute {
  id: Int!
  name: String
}

type TemporaryNotification {
  content: Text
  startDate: String
  endDate: String
}

type Mareograph {
  id: Int!
  name: String
  dateTime: Float
  waterLevel: Float
  n2000WaterLevel: Float
  calculated: Boolean
}
