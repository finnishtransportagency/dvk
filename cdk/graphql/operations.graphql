query findFairways {
  fairways {
    ...fragment FairwayFragment on Fairway {
      id
      nameFI
      nameSV
    }
  }
}

query findFairway($id: Int) {
  fairway(id: $id) {
    ...Fairway
  }
}
