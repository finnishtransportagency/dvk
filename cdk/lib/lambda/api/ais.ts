import { Feature, FeatureCollection, GeoJsonProperties, Geometry } from 'geojson';
import { Vessel, VesselAPIModel, VesselLocationFeatureCollection } from './apiModels';
import { fetchAISFeatureCollection, fetchAISMetadata } from './axios';

const aisV1Path = 'ais/api/ais/v1/';

function getTimestampParams(): Record<string, string> {
  const now = Date.now();
  const yesterday = Date.now() - 24 * 60 * 60 * 1000;
  return { from: yesterday.toString(), to: now.toString() };
}

function parseFrom64BitInteger(integer: number, bitStart: number, bitEnd: number) {
  // 64-bit integer: get bit range from start to end by losing other bits
  // Shift bits first to left and then to right, let others fall off
  const i = (integer << (63 - bitStart)) >>> (63 - bitStart + bitEnd);
  return i.toString().padStart(2, '0');
}

/* MMDDHHMM UTC
Bits 19-16: month; 1-12; 0 = not available = default
Bits 15-11: day; 1-31; 0 = not available = default
Bits 10-6: hour; 0-23; 24 = not available = default
Bits 5-0: minute; 0-59; 60 = not available = default
*/
function parseETA(eta: number): string {
  const month = parseFrom64BitInteger(eta, 19, 16);
  const day = parseFrom64BitInteger(eta, 15, 11);
  const hour = parseFrom64BitInteger(eta, 10, 6);
  const minute = parseFrom64BitInteger(eta, 5, 0);
  return month + '-' + day + ' ' + hour + ':' + minute;
}

export async function fetchVessels(): Promise<Vessel[]> {
  const path = aisV1Path + 'vessels';
  const params = getTimestampParams();
  const data: VesselAPIModel[] = await fetchAISMetadata(path, params);
  return data.map((row) => {
    return {
      name: row.name,
      timestamp: new Date(row.timestamp),
      mmsi: row.mmsi,
      callSign: row.callSign,
      imo: row.imo,
      shipType: row.shipType,
      draught: row.draught / 10,
      eta: parseETA(row.eta),
      posType: row.posType,
      referencePointA: row.referencePointA,
      referencePointB: row.referencePointB,
      referencePointC: row.referencePointC,
      referencePointD: row.referencePointD,
      destination: row.destination,
    };
  });
}

export async function fetchLocations(): Promise<FeatureCollection> {
  const path = aisV1Path + 'locations';
  const params = getTimestampParams();
  const data: VesselLocationFeatureCollection = await fetchAISFeatureCollection(path, params);
  const { dataUpdatedTime, features } = data;
  const geojsonFeatures: Feature<Geometry, GeoJsonProperties>[] = features.map((f) => {
    return {
      type: 'Feature',
      id: f.mmsi,
      geometry: f.geometry,
      properties: {
        ...f.properties,
        dataUpdatedTime,
        timestampExternal: new Date(f.properties.timestampExternal),
        featureType: 'aisvessel',
      },
    };
  });
  const collection: FeatureCollection = {
    type: 'FeatureCollection',
    features: geojsonFeatures,
  };
  return collection;
}
