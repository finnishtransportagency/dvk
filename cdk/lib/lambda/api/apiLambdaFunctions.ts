import * as path from 'path';

interface BackendAPILambda {
  entry: string;
  pathPattern: string;
  functionName: string;
  priority: number;
  useVpc?: boolean;
  useMonitoring?: boolean;
}

const apiLambdaFunctions: BackendAPILambda[] = [
  {
    entry: path.join(__dirname, 'featureloader-handler.ts'),
    pathPattern: '/api/featureloader',
    functionName: 'featureloader',
    priority: 10,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'aislocations-handler.ts'),
    pathPattern: '/api/aislocations',
    functionName: 'aislocations',
    priority: 14,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'aisvessels-handler.ts'),
    pathPattern: '/api/aisvessels',
    functionName: 'aisvessels',
    priority: 16,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'pilotroute-handler.ts'),
    pathPattern: '/api/pilotroutes',
    functionName: 'pilotroutes',
    priority: 18,
    useVpc: true,
    useMonitoring: true,
  },
  {
    entry: path.join(__dirname, 'login.ts'),
    pathPattern: '/yllapito/kirjaudu.html',
    functionName: 'login',
    priority: 20,
    useVpc: false,
  },
  {
    entry: path.join(__dirname, 'logout.ts'),
    pathPattern: '/yllapito/api/logout',
    functionName: 'logout',
    priority: 30,
    useVpc: false,
  },
];

export default apiLambdaFunctions;
