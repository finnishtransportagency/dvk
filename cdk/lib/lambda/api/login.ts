import { getCloudFrontPrivateKey, getCloudFrontPublicKeyId, getCognitoUrl } from '../environment';
import { getSignedCookies } from '@aws-sdk/cloudfront-signer';
import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import JWT, { JwtPayload } from 'jsonwebtoken';
import jwkToPem from 'jwk-to-pem';
import axios from 'axios';

const cloudFrontDnsName = process.env.CLOUDFRONT_DNSNAME;
let cachedKeys: Record<string, string>;

// Fetch JWK's from Cognito or cache
const getPublicKeys = async (issuerUrl: string) => {
  if (!cachedKeys) {
    cachedKeys = {};
    const publicKeys = await axios.get(issuerUrl + '/.well-known/jwks.json');
    for (const key of publicKeys.data.keys) {
      cachedKeys[key.kid] = jwkToPem(key);
    }
    return cachedKeys;
  } else {
    return cachedKeys;
  }
};

export const validateJwtToken = async (token: string | undefined, dataToken: string): Promise<JwtPayload | undefined> => {
  if (!token) {
    log.debug('IAM JWT Token missing');
    return;
  }
  // Split token into parts
  const tokenParts = token.split('.');
  if (tokenParts.length < 2) {
    log.error('Invalid token');
    return;
  }

  // Parse header & payload from token parts
  const tokenHeader = JSON.parse(Buffer.from(tokenParts[0], 'base64').toString('utf-8'));
  const tokenPayload = JSON.parse(Buffer.from(tokenParts[1], 'base64').toString('utf-8'));

  // Fetch public keys from Cognito
  const publicKeys = await getPublicKeys(tokenPayload.iss);
  const publicKey = publicKeys[tokenHeader.kid];
  if (!publicKey) {
    log.debug('Public key not found');
    return;
  }

  // Verify token
  const result = JWT.verify(token, publicKey, { issuer: await getCognitoUrl() }) as JwtPayload;
  if (!result) {
    log.error('Failed to verify JWT');
    return;
  }

  // Check use access
  if (result.token_use !== 'access') {
    log.error('Invalid token use');
    return;
  }

  // Return decoded data token
  return JWT.decode(dataToken.replace(/=/g, '')) as JwtPayload;
};

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  try {
    let jwtToken;
    if (event.multiValueHeaders) {
      log.info({ headers: event.multiValueHeaders }, 'Request headers');
      const token = event.multiValueHeaders['x-iam-accesstoken'];
      const data = event.multiValueHeaders['x-iam-data'];
      if (token && data) {
        jwtToken = await validateJwtToken(token[0], data[0]);
      }
    }
    log.info('JwtToken: %o', jwtToken);
    // TODO: comment out once token received from ALB
    /*if (!jwtToken) {
      return {
        statusCode: 403,
      };
    }*/
    const cloudFrontPolicy = JSON.stringify({
      Statement: [
        {
          Resource: `https://${cloudFrontDnsName}/*`,
          Condition: {
            DateLessThan: {
              'AWS:EpochTime': Math.floor(new Date().getTime() / 1000) + 3600,
            },
          },
        },
      ],
    });

    const cookie = getSignedCookies({
      url: cloudFrontDnsName as string,
      privateKey: await getCloudFrontPrivateKey(),
      keyPairId: await getCloudFrontPublicKeyId(),
      policy: cloudFrontPolicy,
    });

    const setCookieAttributes = `; Domain=${cloudFrontDnsName}; Path=/; Secure; HttpOnly; SameSite=Lax`;

    const response: ALBResult = {
      statusCode: 302,
      multiValueHeaders: {
        Location: [`https://${cloudFrontDnsName}/yllapito/index.html`],
        'Set-Cookie': [
          `CloudFront-Key-Pair-Id=${cookie['CloudFront-Key-Pair-Id']}${setCookieAttributes}`,
          `CloudFront-Policy=${cookie['CloudFront-Policy']}${setCookieAttributes}`,
          `CloudFront-Signature=${cookie['CloudFront-Signature']}${setCookieAttributes}`,
        ],
      },
    };
    log.debug(response, 'Login response');
    return response;
  } catch (e) {
    log.error('Getting signed cookies failed: %s', e);
    return { statusCode: 500, body: String(e) };
  }
};
