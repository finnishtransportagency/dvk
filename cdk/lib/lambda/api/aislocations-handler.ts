import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchLocations } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aislocations()`);
  return handleAisCall('aislocations', fetchLocations, ['application/geo+json']);
};
