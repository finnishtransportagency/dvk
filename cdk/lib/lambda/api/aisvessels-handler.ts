import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchVessels } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aisvessels()`);
  return handleAisCall('aisvessels', fetchVessels, ['application/json']);
};
