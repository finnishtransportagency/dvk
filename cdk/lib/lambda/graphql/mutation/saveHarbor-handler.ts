import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { Harbor, HarborInput, Maybe, MutationSaveHarborArgs, TextInput } from '../../../../graphql/generated';
import { log } from '../../logger';
import HarborDBModel from '../../db/harborDBModel';
import { getCurrentUser } from '../../api/login';

function mapText(text?: Maybe<TextInput>) {
  if (text) {
    return {
      fi: text.fi,
      sv: text.sv,
      en: text.en,
    };
  } else {
    return undefined;
  }
}

function map<T>(text: Maybe<T> | undefined): T | undefined {
  return text ? text : undefined;
}

function mapStringArray(text: Maybe<Maybe<string>[]> | undefined): string[] | undefined {
  return text ? (text.map((t) => map<string>(t)).filter((t) => t !== undefined) as string[]) : undefined;
}

function mapHarborToModel(harbor: HarborInput, old: HarborDBModel | undefined): HarborDBModel {
  return {
    id: harbor.id,
    name: harbor.name,
    status: harbor.status,
    company: mapText(harbor.company),
    creator: old ? old.creator : 'Erkki Esimerkki',
    creationTimestamp: old ? old.creationTimestamp : Date.now(),
    modifier: 'Erkki Esimerkki',
    modificationTimestamp: Date.now(),
    email: map<string>(harbor.email),
    extraInfo: mapText(harbor.extraInfo),
    fax: map<string>(harbor.fax),
    harborBasin: mapText(harbor.harborBasin),
    internet: map<string>(harbor.internet),
    phoneNumber: mapStringArray(harbor.phoneNumber),
    geometry: harbor.geometry ? { type: 'Point', coordinates: [harbor.geometry.lat, harbor.geometry.lon] } : undefined,
    cargo: harbor.cargo
      ?.map((c) => {
        return { fi: c?.fi || undefined, sv: c?.sv || undefined, en: c?.en || undefined };
      })
      .filter((c) => c.fi || c.sv || c.en),
    quays: harbor.quays?.map((q) => {
      return {
        extraInfo: mapText(q?.extraInfo),
        length: map<number>(q?.length),
        name: mapText(q?.name),
        geometry: q?.geometry ? { type: 'Point', coordinates: [q.geometry.lat, q.geometry.lon] } : undefined,
        sections: q?.sections?.map((s) => {
          return {
            depth: map<number>(s?.depth),
            name: map<string>(s?.name),
            geometry: s?.geometry ? { type: 'Point', coordinates: [s.geometry.lat, s.geometry.lon] } : undefined,
          };
        }),
      };
    }),
  };
}

export const handler: AppSyncResolverHandler<MutationSaveHarborArgs, Harbor> = async (
  event: AppSyncResolverEvent<MutationSaveHarborArgs>
): Promise<Harbor> => {
  const user = await getCurrentUser(event);
  log.info(`saveHarbor(${event.arguments.harbor?.id}, ${user.uid})`);
  if (event.arguments.harbor?.id) {
    const dbModel = await HarborDBModel.get(event.arguments.harbor.id);
    const newModel = mapHarborToModel(event.arguments.harbor, dbModel);
    log.debug('harbor: %o', newModel);
    await HarborDBModel.save(newModel);
    return newModel;
  }
  throw new Error('Harbor id missing');
};
