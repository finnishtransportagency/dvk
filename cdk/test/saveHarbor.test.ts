import { HarborInput, Operation, OperationError, Status } from '../graphql/generated';
import { CurrentUser } from '../lib/lambda/api/login';
import { mapHarborToModel } from '../lib/lambda/graphql/mutation/saveHarbor-handler';

const currentUser: CurrentUser = { firstName: 'Test', lastName: 'Test', uid: 'test', roles: [] };
test('test mapHarborToModel minimal', () => {
  expect(() => mapHarborToModel({} as HarborInput, undefined, currentUser)).toThrow(OperationError.InvalidInput);
  const input: HarborInput = {
    id: 'a1',
    version: 'v2',
    name: { fi: 'test', sv: 'testsv', en: 'testen' },
    n2000HeightSystem: false,
    operation: Operation.Create,
    geometry: { lat: '60', lon: '20' },
    status: Status.Draft,
  };
  expect(mapHarborToModel(input, undefined, currentUser)).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});

test('test mapHarborToModel all', () => {
  const input: HarborInput = {
    id: 'a2',
    version: 'v2',
    name: { fi: 'test2', sv: 'testsv2', en: 'testen2' },
    n2000HeightSystem: true,
    operation: Operation.Update,
    geometry: { lat: '62', lon: '18' },
    status: Status.Public,
    cargo: {
      fi: 'cargofi',
      sv: 'cargosv',
      en: 'cargoen',
    },
    company: {
      fi: 'companyfi',
      sv: 'companysv',
      en: 'companyen',
    },
    email: 'email@fi',
    extraInfo: {
      fi: 'extrafi',
      sv: 'extrasv',
      en: 'extraen',
    },
    fax: '32321',
    internet: 'www.fi',
    phoneNumber: ['12345', '22222'],
    harborBasin: {
      fi: 'basinfi',
      sv: 'basinsv',
      en: 'basinen',
    },
    quays: [
      {
        extraInfo: {
          fi: 'cargofi',
          sv: 'cargosv',
          en: 'cargoen',
        },
        name: {
          fi: 'cargofi',
          sv: 'cargosv',
          en: 'cargoen',
        },
        geometry: { lat: '63.2', lon: '22.1' },
        length: '123',
        sections: [
          {
            depth: '10',
            geometry: { lat: '62.2', lon: '17' },
            name: 'stest',
          },
        ],
      },
    ],
  };
  expect(mapHarborToModel(input, undefined, currentUser)).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});
