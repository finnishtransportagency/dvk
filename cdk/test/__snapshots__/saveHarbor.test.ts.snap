// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test mapHarborToModel all 1`] = `
{
  "cargo": {
    "en": "cargoen",
    "fi": "cargofi",
    "sv": "cargosv",
  },
  "company": {
    "en": "companyen",
    "fi": "companyfi",
    "sv": "companysv",
  },
  "creationTimestamp": Any<Number>,
  "creator": "Test Test",
  "email": "email@fi",
  "expires": null,
  "extraInfo": {
    "en": "extraen",
    "fi": "extrafi",
    "sv": "extrasv",
  },
  "fax": "32321",
  "geometry": {
    "coordinates": [
      18,
      62,
    ],
    "type": "Point",
  },
  "harborBasin": {
    "en": "basinen",
    "fi": "basinfi",
    "sv": "basinsv",
  },
  "id": "a2",
  "internet": "www.fi",
  "modificationTimestamp": Any<Number>,
  "modifier": "Test Test",
  "n2000HeightSystem": true,
  "name": {
    "en": "testen2",
    "fi": "test2",
    "sv": "testsv2",
  },
  "phoneNumber": [
    "12345",
    "22222",
  ],
  "publishDetails": undefined,
  "quays": [
    {
      "extraInfo": {
        "en": "cargoen",
        "fi": "cargofi",
        "sv": "cargosv",
      },
      "geometry": {
        "coordinates": [
          22.1,
          63.2,
        ],
        "type": "Point",
      },
      "length": 123,
      "name": {
        "en": "cargoen",
        "fi": "cargofi",
        "sv": "cargosv",
      },
      "sections": [
        {
          "depth": 10,
          "geometry": {
            "coordinates": [
              17,
              62.2,
            ],
            "type": "Point",
          },
          "name": "stest",
        },
      ],
    },
  ],
  "status": "PUBLIC",
  "version": "v2",
}
`;

exports[`test mapHarborToModel minimal 1`] = `
{
  "cargo": null,
  "company": null,
  "creationTimestamp": Any<Number>,
  "creator": "Test Test",
  "email": null,
  "expires": null,
  "extraInfo": null,
  "fax": null,
  "geometry": {
    "coordinates": [
      20,
      60,
    ],
    "type": "Point",
  },
  "harborBasin": null,
  "id": "a1",
  "internet": null,
  "modificationTimestamp": Any<Number>,
  "modifier": "Test Test",
  "n2000HeightSystem": false,
  "name": {
    "en": "testen",
    "fi": "test",
    "sv": "testsv",
  },
  "phoneNumber": null,
  "publishDetails": undefined,
  "quays": null,
  "status": "DRAFT",
  "version": "v2",
}
`;
